### ISSUE TITLE
Add support for vite.config.ts file

### ISSUE THREAD

**Issue by EnvyIT** [smapiot/piral-cli-vite#7]
# Support for TypeScript configuration files

## Description
Currently, it seems that only vite.config.js files are supported. But it would be great if vite.config.ts files could be used as well. 

## Discussion

The best argument for me is, that [vitejs](https://vitejs.dev/config/ ) does support this out of the box and it is an expected behavior if one creates a Piral project using TS + Vite to have support for config files in TS too.

**Comment by FlorianRappl** [comment]
Yes, I think we can do that. Incoming.

**Comment by FlorianRappl** [comment]
Supported with 1.1.0

**Comment by KonstantinBozhkov** [comment]
When I use version 1.1.0, the vite.config.(js|ts) file is ignored. However, when I use version 1.0.3, vite.config.js works.
@FlorianRappl, can you double-check your changes?

**Comment by FlorianRappl** [comment]
> When I use version 1.1.0, the vite.config.(js|ts) file is ignored. However, when I use version 1.0.3, vite.config.js works.

Hm can you specify what you mean with "ignored". Because the format of file changed; it is now the Vite format - not just an extension / config manipulator. What exactly are you trying?

**Comment by EnvyIT** [comment]
Hi, I can confirm that the vite.config.ts|js files are ignored now on running `piral build` command.

This is an example of my configuration `vite.config.js` with version `1.0.3` which worked perfectly fine:

```javascript
const { resolve } = require('path');

const publicDir = () => resolve(__dirname, 'public');
const atPath = () => resolve(__dirname, 'src/');
const apiPath = () => `${resolve(__dirname, '/src/api/')}`;
const corePath = () => `${resolve(__dirname, '/src/core/')}`;
const featurePath = () => `${resolve(__dirname, '/src/feature/')}`;

const onWarningHandler = (warning, warn) => {
  if ('MODULE_LEVEL_DIRECTIVE' === warning.code) {
    return;
  }
  warn(warning);
};

const extendedConfig = {
  root: './src',
  resolve: {
    alias: {
      '@': atPath(),
      api: apiPath(),
      core: corePath(),
      feature: featurePath(),
    },
  },
  publicDir: publicDir(),
  server: {
    port: 3000,
    proxy: {},
  },
  test: {
    globals: true,
    restoreMocks: true,
    unstubGlobals: true,
    environment: 'jsdom',
    setupFiles: ['./__test__/setup.ts'],
    watch: false,
    maxConcurrency: 20,
    parallel: true,
    useAtomics: true,
    experimentalVmThreads: true,
  },
};

module.exports = function(config) {
  config.build.target = 'esnext';
  config.build.emptyOutDir = true;
  config.build.rollupOptions.onwarn = onWarningHandler;
  return { ...config, ...extendedConfig };
};
```

with the update to version `1.1.0` the test execution works perfectly fine with vitest where I pass my configuration file directly.

Unfortunately, when I exectue the build I get the following error for my resolve aliases:

```
🦋  piral-cli v1.5.0
ℹ Bundle release ...
vite v4.5.2 building for production...
transforming...
✓ 8 modules transformed.
✓ built in 219ms
[vite]: Rollup failed to resolve import "@/core/config" from "/home/meder/projects/piral/src/App.tsx".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
✖ [0174] The bundling process failed: [vite]: Rollup failed to resolve import "@/core/config" from "/home/meder/piral/src/App.tsx".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
🐛  Find detailed descriptions at https://docs.piral.io/code/search
```

and when they are "pruned" and the ugly imports are used the warning handler is not working and I get a bunch of warnings during the build. (e.g.:

```
node_modules/@mui/material/Stepper/Stepper.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/Stepper/Stepper.js" was ignored.
node_modules/@mui/material/SvgIcon/SvgIcon.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/SvgIcon/SvgIcon.js" was ignored.
node_modules/@mui/material/Switch/Switch.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/Switch/Switch.js" was ignored.
node_modules/@mui/material/SwipeableDrawer/SwipeableDrawer.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/SwipeableDrawer/SwipeableDrawer.js" was ignored.
node_modules/@mui/material/Tab/Tab.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/Tab/Tab.js" was ignored.
node_modules/@mui/material/Table/Table.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/Table/Table.js" was ignored.
node_modules/@mui/material/TableBody/TableBody.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableBody/TableBody.js" was ignored.
node_modules/@mui/material/TableCell/TableCell.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableCell/TableCell.js" was ignored.
node_modules/@mui/material/TableContainer/TableContainer.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableContainer/TableContainer.js" was ignored.
node_modules/@mui/material/TableFooter/TableFooter.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableFooter/TableFooter.js" was ignored.
node_modules/@mui/material/TableHead/TableHead.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableHead/TableHead.js" was ignored.
node_modules/@mui/material/TablePagination/TablePagination.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TablePagination/TablePagination.js" was ignored.
node_modules/@mui/material/TableRow/TableRow.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableRow/TableRow.js" was ignored.
node_modules/@mui/material/TableSortLabel/TableSortLabel.js (1:0) Module level directives cause errors when bundled, "use client" in "node_modules/@mui/material/TableSortLabel/TableSortLabel.js" was ignored.
```

Now neither the `js` nor the `ts` version, which looks way different than this customized js config version explicitly for piral, is working.

Would be glad if you could give me some insights what I am doing wrong (?)
I lost more time during my PoC on configuration issues from a React App with vite than I initially thought.

As a side note: the version of vite is also already bit oudate since today version `5.1.4` is already available and the project still runs on `4.5.2`. We are using `renovate` in production to update to latest versions as soon as they are stable but it seems with piral this is no longer possible and we are dependent on your release cycles.

**Comment by FlorianRappl** [comment]
> We are using renovate in production to update to latest versions as soon as they are stable but it seems with piral this is no longer possible and we are dependent on your release cycles.

Well, if you want to take `piral-cli-vite` this is true. But then nothing prevents you using your own bundler. The idea behind the bundler plugins is that you get everything out of the box - nothing you need to worry about.

Also if you want to influence release cycles / where effort is spend (e.g., maintaining `piral-cli-vite`) I'd recommend working together. We got some sponsorship possibilities and other ways of working together that might come in handy.

**Comment by EnvyIT** [comment]
First of all, thanks for your quick response @FlorianRappl! I really appreciate your work on the framework.

To your point with the custom builder: To be honest, I tried the "easy" way first if it is working out of the box, but you are totally right if we need more control we either need to contribute or write our own builder.

Back to my first question: Any idea why I get this issue with the `piral build` step where it seems that the configuration is ignored?

**Comment by FlorianRappl** [comment]
I think here we have two cases. First case:

```plain
$ npm init pilet -- --source sample-piral --bundler vite --defaults
✔ Pilet scaffolded successfully!
$ echo 'console.log("Hi!");' > vite.config.ts
$ more vite.config.ts 
console.log("Hi!");
$ npm run build

> p7@1.0.0 build
> pilet build   

🦋  piral-cli v1.5.0
ℹ Bundle pilet ...
Hi!
vite v4.5.2 building for production...
transforming...
✓ 2 modules transformed.
rendering chunks...   
computing gzip size...
dist/index.js          0.53 kB │ gzip: 0.38 kB │ map: 0.96 kB
dist/Page.97834f1b.js  0.99 kB │ gzip: 0.52 kB │ map: 1.21 kB
✓ built in 73ms
✔ Pilet "p7" built successfully!
```

See the "Hi!" - it is not ignored (`pilet build`).

However, in the second case this looks different:

```plain
$ npm init piral-instance -- --bundler vite --defaults
Need to install the following packages:
create-piral-instance@1.5.0
Ok to proceed? (y) y
✔ Piral instance scaffolded successfully!
$ echo 'console.log("Hi!");' > vite.config.ts
$ npm run build

> as2@1.0.0 build
> piral build    

🦋  piral-cli v1.5.0
ℹ Bundle emulator ...
vite v4.5.2 building for production...
transforming...
✓ 223 modules transformed.
rendering chunks...
computing gzip size...
../dist/emulator/app/index.html              0.50 kB │ gzip:   0.34 kB
../dist/emulator/app/index.61f0f2f4.css      3.84 kB │ gzip:   1.22 kB
../dist/emulator/app/index.82df9d2b.js   1,195.87 kB │ gzip: 216.67 kB │ map: 2,156.88 kB
✓ built in 3.15s
ℹ Bundle release ...
vite v4.5.2 building for production...
transforming...
✓ 205 modules transformed.
rendering chunks...
computing gzip size...
../dist/release/index.html            0.50 kB │ gzip:  0.34 kB
../dist/release/index.c85dc9b3.css    3.25 kB │ gzip:  1.14 kB
../dist/release/index.3e1a4d19.js   246.36 kB │ gzip: 80.59 kB │ map: 806.46 kB
✓ built in 1.85s
✔ Files for publication available in "dist/release".
```

As you can see no `Hi!` here as you've also found out. But wait - Vite always looks in its "root" and the root from Vite is where the `index.html` is placed. So let's move it:

```plain
$ mv vite.config.ts src/vite.config.ts
$ npm run build

> as2@1.0.0 build
> piral build

🦋  piral-cli v1.5.0
ℹ Bundle emulator ...
Hi!
vite v4.5.2 building for production...
transforming...
✓ 223 modules transformed.
rendering chunks...
computing gzip size...
../dist/emulator/app/index.html              0.50 kB │ gzip:   0.34 kB
../dist/emulator/app/index.61f0f2f4.css      3.84 kB │ gzip:   1.22 kB
../dist/emulator/app/index.4f90e886.js   1,195.87 kB │ gzip: 216.67 kB │ map: 2,156.88 kB
✓ built in 3.08s
ℹ Bundle release ...
Hi!
vite v4.5.2 building for production...
transforming...
✓ 205 modules transformed.
rendering chunks...
computing gzip size...
../dist/release/index.html            0.50 kB │ gzip:  0.34 kB
../dist/release/index.c85dc9b3.css    3.25 kB │ gzip:  1.14 kB
../dist/release/index.3e1a4d19.js   246.36 kB │ gzip: 80.59 kB │ map: 806.46 kB
✓ built in 1.93s
✔ Files for publication available in "dist/release".
```

Of course, this is not really wanted (and I've seen that the docs are a bit outdated on that, too), so I'll try to improve the situation by supporting both - the standard Vite location and the expected location.

Regarding Vite 5 - the main reason I did not upgrade is that it might break for users. One way out of this would be to do it the Webpack way, i.e., introduce `piral-cli-vite5`, but personally I don't like it very much. Having so many packages when all you want / need is a way for users to select what version they want (surely, under the hood we might need to do things differently per Vite version, but usually these are edge cases / not many changes).

**Comment by EnvyIT** [comment]
Thanks again for your detailed explaination. I will take a look on it and take further steps in the direction of a custom bundler, cause in the end I prefer to have fully control even when I use a framework.

I totally agree with your opinion regarding the Webpack way. I would not prefer such kind of solution too.
Will keep you up to date and give feedback after following the docs for custom bundlers and implementing it on my own.

Big thanks for your support.

**Comment by FlorianRappl** [comment]
Fix for it is out (1.1.1) -> https://www.npmjs.com/package/piral-cli-vite
